armory/data/datasets.py

```diff
@@ -20,70 +21,98 @@
 logger = logging.getLogger(__name__)
 
 
-def _in_memory_dataset_tfds(
-    dataset_name: str, dataset_dir: str, preprocessing_fn: Callable,
-) -> (np.ndarray, np.ndarray, np.ndarray, np.ndarray):
-    default_graph = tf.compat.v1.keras.backend.get_session().graph
+class ArmoryDataGenerator(DataGenerator):
+    def __init__(self, generator, size, batch_size, preprocessing_fn=None):
+        super().__init__(size, batch_size)
+        self.preprocessing_fn = preprocessing_fn
+        self.generator = generator
+        self.total_iterations = size // batch_size
 
-    train_x, train_y = tfds.load(
-        dataset_name,
-        batch_size=-1,
-        split="train",
-        as_supervised=True,
-        data_dir=dataset_dir,
-    )
+    def get_batch(self) -> (np.ndarray, np.ndarray):
+        x, y = next(self.generator)
+        if self.preprocessing_fn:
+            x = self.preprocessing_fn(x)
+
+        return x, y
+
+
+def _generator_from_tfds(
+    dataset_name: str,
+    split_type: str,
+    batch_size: int,
+    epochs: int,
+    dataset_dir: str,
+    preprocessing_fn: Callable,
+):
+    default_graph = tf.compat.v1.keras.backend.get_session().graph
 
-    test_x, test_y = tfds.load(
+    ds, ds_info = tfds.load(
         dataset_name,
-        batch_size=-1,
-        split="test",
+        batch_size=batch_size,
+        split=split_type,
         as_supervised=True,
```

Comment:
Not all of our datasets allow as_supervised=True

Reply:
Ah thats a good catch. Can we change this when we add dataset that requires this (I think resisc45?)

It's a lot easier to analyze/work through when we have a dataset we can test with.